### PR TITLE
types: add error types

### PR DESCRIPTION
**Pull request by niedbalski**
# Summary of this proposal

Error types must be part of the shared types package, also its corresponding string transformation
methods.

## Notes for the reviewer

N/A

## Issue(s) number(s)

N/A

## Checklist for submitter

- [X] My PR requires updating dependencies
- [X] My PR has the documentation changes required.

## Checklist for reviewer

- [ ] The proposal fixes a bug/issue or implements a new feature that is well described.
- [ ] The proposal has sufficient test cases that covers the changes.
  - [ ] If changes an API, it does not break backwards compatibility (-1 major version).
- [ ] If integration is required, the proposal has integration tests.
- [ ] The proposal does not break coverage stats
- [ ] The proposal has the required documentation changes.

## Backport

- [ ] Backport to the latest stable release.

### CODE REVIEW

File: types/agent.go [hunk 1]
```diff
@@ -6,6 +6,29 @@ import (
...
+	ErrInvalidAgentRawConfig         = InvalidArgumentError("invalid agent raw config")
+	ErrInvalidAgentMetadata          = InvalidArgumentError("invalid agent metadata")
+	ErrUpdateAgentVersionForbidden   = PermissionDeniedError("update agent version forbidden")
+	ErrUpdateAgentEditionForbidden   = PermissionDeniedError("update agent edition forbidden")
+	ErrUpdateAgentFlagsForbidden     = PermissionDeniedError("update agent flags forbidden")
+	ErrUpdateAgentRawConfigForbidden = PermissionDeniedError("update agent raw config forbidden")
+	ErrUpdateAgentMetadataForbidden  = PermissionDeniedError("update agent metadata forbidden")
+	ErrAgentGone                     = GoneError("agent gone")
+	ErrMachineIDConflict             = ConflictError("machine ID conflict")
+	ErrAgentNotFound                 = NotFoundError("agent not found")
```
Comment (patrick-stephens) [inline types/agent.go:29]: Not found but not gone? :-)
Permanently deleted might be better text?

File: types/pipeline_config.go [hunk 2]
```diff
@@ -2,6 +2,13 @@ package types
 
 import "time"
 
+const (
+	ErrInvalidPipelineConfigID = InvalidArgumentError("invalid pipeline config ID")
+	ErrInvalidPipelineConfig   = InvalidArgumentError("invalid pipeline config")
+	ErrZeroPipelineConfigPorts = InvalidArgumentError("zero pipeline config ports")
+	ErrPipelineConfigNotFound  = NotFoundError("pipeline config not found")
```
Comment (patrick-stephens) [inline types/pipeline_config.go:9]: No Gone error?

File: types/user.go [hunk 3]
```diff
@@ -2,6 +2,15 @@ package types
 
 import "time"
 
+const (
+	ErrEmailNotVerified = UnauthenticatedError("email not verified")
+	ErrInvalidEmail     = InvalidArgumentError("invalid email")
+	ErrInvalidUserName  = InvalidArgumentError("invalid user name")
+	ErrInvalidAvatarURL = InvalidArgumentError("invalid avatar URL")
+	ErrEmailTaken       = ConflictError("email taken")
```
Comment (patrick-stephens) [inline types/user.go:10]: Do we want to expose this? It leaks account credentials potentially versus just an "invalid email".

Comment (niedbalski) [reply types/user.go:10]: I think I agree, lets remove it from the public and fix the API to don't leak this specific error.

File: types/pipeline_config.go (hunk 2 above)
Comment (niedbalski) [reply types/pipeline_config.go:9]: Not in use now, but we might add it.

**Comment by niedbalski** [comment]
Not following up this change.

**Comment by nicolasparada** [comment]
🙇‍♂️

**Pull request closed without merging**